Add top level docs for tracing appender (#692)
tokio-rs/tracing: tracing-appender/README.md

[![Discord chat][discord-badge]][discord-url]

[Documentation][docs-url] | [Chat][discord-url]

[crates-badge]: https://img.shields.io/crates/v/tracing-appender.svg
[crates-url]: https://crates.io/crates/tracing-appender/0.1.0
[docs-badge]: https://docs.rs/tracing-appender/badge.svg
[docs-url]: https://docs.rs/tracing-appender/0.1.0
[docs-master-badge]: https://img.shields.io/badge/docs-master-blue
[docs-master-url]: https://tracing-rs.netlify.com/tracing-appender
[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg
[mit-url]: LICENSE
[actions-badge]: https://github.com/tokio-rs/tracing/workflows/CI/badge.svg
[actions-url]:https://github.com/tokio-rs/tracing/actions?query=workflow%3ACI
[discord-badge]: https://img.shields.io/discord/500028886025895936?logo=discord&label=discord&logoColor=white
[discord-url]: https://discord.gg/EeF3cQw

## Overview

[`tracing`] is a framework for instrumenting Rust programs to collect
structured, event-based diagnostic information. This crate provides the ability
for [`tracing`] events and spans to be recorded in a non-blocking manner by using a
dedicated logging thread. It also provides a [`RollingFileAppender`]

> that can be used with or without the non-blocking writer.

Author: I don't think you meant to cut off "[`tracing`] is a framework for " otherwise we now start off with 

"structured, event-based"

## Usage

First, add this to your `Cargo.toml`:
```toml
tracing-appender = "0.1"
```

This crate can be used in a few ways to record spans/events:
 - Using a [`RollingFileAppender`] to perform writes to a log file. This will block on writes.

>  - Using *any* `std::io::Write` implementation in a non-blocking way

Reviewer: ```suggestion
 - Using *any* `std::io::Write` implementation in a non-blocking way.
```

 - Using a combination of [`NonBlocking`] and [`RollingFileAppender`] to allow writes to a log file
without blocking.

## Rolling File Appender

```rust
fn main(){
    let file_appender = tracing_appender::rolling::hourly("/some/directory", "prefix.log");
}
```
This creates an hourly rotating file appender which outputs to `/some/directory/prefix.log.YYYY-MM-DD-HH`.
[`Rotation::DAILY`] and [`Rotation::NEVER`] are the other available options.

It implements the `std::io::Write` trait. To use with a subscriber, it must be combined with a
[`MakeWriter`] implementation to be able to record tracing spans/event.

The [rolling] module's documentation provides more detail on how to use this file appender.

## Non-Blocking Writer

> 

Reviewer: ```suggestion
The example below demonstrates the construction of a `non_blocking` writer with `std::io::stdout()`, which implements `MakeWriter`.
```

```rust
fn main() {
    let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::stdout());
    let subscriber = tracing_subscriber::fmt().with_writer(non_blocking);
    tracing::subscriber::set_global_default(subscriber.finish()).expect("Could not set global default");
}
```
**Note:** `_guard` is a [`WorkerGuard`] which is returned by `tracing_appender::non_blocking`
to ensure buffered logs are flushed to their output in the case of abrupt terminations of a process.
See [`WorkerGuard`] module for more details.

> Alternatively, you can provide *any* type that implements [`std::io::Write`] to
> [`tracing_appender::non_blocking`]

Reviewer: ```suggestion
The example below demonstrates the construction of a [`tracing_appender::non_blocking`] writer constructed with a [`std::io::Write`]:
```

(Provide an example below. Copy and pasting the example above with a `MakeWriter` but replacing the `std::io::stdout()` with, for instance, a [`std::io::Stdout`](https://doc.rust-lang.org/std/io/struct.Stdout.html) would be good.

Author: I've added an example containing a struct which implements `std::io::Write`

# tracing-appender

Writers for logging events and spans

[![Crates.io][crates-badge]][crates-url]
[![Documentation][docs-badge]][docs-url]
[![Documentation (master)][docs-master-badge]][docs-master-url]
[![MIT licensed][mit-badge]][mit-url]
[![Build Status][actions-badge]][actions-url]
[![Discord chat][discord-badge]][discord-url]

[Documentation][docs-url] | [Chat][discord-url]

[crates-badge]: https://img.shields.io/crates/v/tracing-appender.svg
[crates-url]: https://crates.io/crates/tracing-appender/0.1.0
[docs-badge]: https://docs.rs/tracing-appender/badge.svg
[docs-url]: https://docs.rs/tracing-appender/0.1.0
[docs-master-badge]: https://img.shields.io/badge/docs-master-blue

> [docs-master-url]: https://tracing-rs.netlify.com/tracing-appender

Reviewer 2: Nit/TIOLI: I believe that we own the domain name `tracing.rs`
```suggestion
[docs-master-url]: https://tracing.rs/tracing-appender
```

[mit-badge]: https://img.shields.io/badge/license-MIT-blue.svg

> [mit-url]: LICENSE

Reviewer 2: This link won't work, since there isn't a `LICENSE` file in this directory. I believe this needs to be 
```suggestion
[mit-url]: ../LICENSE
```
(but don't quote me on this)

Author: Oh I see, I had copied an pasted from one of the other crates which used ../LICENSE. I see that it also exists at the root dir.

## Usage

Add the following to your `Cargo.toml`:
```toml
tracing-appender = "0.1"
```

This crate can be used in a few ways to record spans/events:
 - Using a [`RollingFileAppender`][file_appender] to write to a log file. This is a blocking operation.
 - Using *any* type implementing [`std::io::Write`][write] in a non-blocking fashion.
 - Using [`NonBlocking`][non_blocking] and [`RollingFileAppender`][file_appender] together to write to write to log files in a non-blocking fashion.

## Rolling File Appender

```rust
fn main(){
    let file_appender = tracing_appender::rolling::hourly("/some/directory", "prefix.log");
}
```
This creates an hourly rotating file appender that writes to `/some/directory/prefix.log.YYYY-MM-DD-HH`.
[`Rotation::DAILY`] and [`Rotation::NEVER`] are the other available options.

The file appender implements [`std::io::Write`][write]. To be used with [`tracing_subscriber::FmtSubscriber`][fmt_subscriber], 
it must be combined with a [`MakeWriter`][make_writer] implementation to be able to record tracing spans/event.

The [rolling] module's documentation provides more detail on how to use this file appender.

## Non-Blocking Writer

> The example below demonstrates the construction of a `non_blocking` writer with `std::io::stdout()`, which implements [`MakeWriter`][[make_writer].

Reviewer 2: This example no longer uses `io::stdout`, can we update this?

Author: Sure

```rust
use std::io::Error;

struct TestWriter;

impl std::io::Write for TestWriter {
    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
        let buf_len = buf.len();
    
        println!("{:?}", buf);
        Ok(buf_len)
    }

    fn flush(&mut self) -> std::io::Result<()> {
        Ok(())
    }
}

fn main() {
    let (non_blocking, _guard) = tracing_appender::non_blocking(TestWriter);
    let subscriber = tracing_subscriber::fmt().with_writer(non_blocking);
    tracing::subscriber::set_global_default(subscriber.finish()).expect("Could not set global default");
}
```
**Note:** `_guard` is a [`WorkerGuard`][guard] which is returned by `tracing_appender::non_blocking`
to ensure buffered logs are flushed to their output in the case of abrupt terminations of a process.
See [`WorkerGuard`][guard] module for more details.

The example below demonstrates the construction of a [`tracing_appender::non_blocking`][non_blocking] writer constructed with a [`std::io::Write`][write]:

> ```rust
> fn main() {
>     let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::Stdout);
>     let subscriber = tracing_subscriber::fmt().with_writer(non_blocking);
>     tracing::subscriber::set_global_default(subscriber.finish()).expect("Could not set global default");
> }

Reviewer 2: TIOLI: What do you think about using the chaining-based API for setting the subscriber?
```suggestion
```rust
fn main() {
    let (non_blocking, _guard) = tracing_appender::non_blocking(std::io::Stdout);
    tracing_subscriber::fmt()
        .with_writer(non_blocking)
        .init();
}
```

```

The [non_blocking] module's documentation provides more detail on how to use `non_blocking`.

## Non-Blocking Rolling File Appender

```rust
fn main() {
    let file_appender = tracing_appender::rolling::hourly("/some/directory", "prefix.log");
    let (non_blocking, _guard) = tracing_appender::non_blocking(file_appender);

>     let subscriber = tracing_subscriber::fmt().with_writer(non_blocking);
>     
>     tracing::subscriber::set_global_default(subscriber.finish()).expect("Could not set global default");

Reviewer 2: similarly, this could be
```suggestion
   tracing_subscriber::fmt()
       .with_writer(non_blocking)
       .init();
```

## Overview

> structured, event-based diagnostic information. `tracing-appender` allows

Author: fixed
